### PR TITLE
Add datacenter node setup scripts

### PR DESCRIPTION
**Pull request by TristanDebrunner**
#### Problem

We need to consistently setup new nodes for the datacenter

#### Summary of Changes

Add bash scripts that do all the setup for us. Also add helper scripts that do parts of the setup if something breaks.

### CODE REVIEW

File: net/datacenter-node-install/scripts/set-hostname.sh [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
+
+if [[ "$EUID" -ne 0 ]]
+  then echo "Please run as root"
+  exit
+fi
+
+set -xe
+
+sed -i 's/preserve_hostname: false/preserve_hostname: true/' /etc/cloud/cloud.cfg
```
Comment (t-nelson) [inline net/datacenter-node-install/scripts/set-hostname.sh]: This should be done via a drop-in to avoid the inevitable merge conflict during `apt upgrade` when vendor defaults change.  Something like
```
echo "preserve_hostname: false" > /etc/cloud.cfg.d/99-disable-preserve-hostname.cfg
```

File: net/datacenter-node-install/scripts/disable-nouveau.sh [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env bash
+
+cat <<EOF > /etc/modprobe.d/blacklist-nouveau.conf
+blacklist nouveau
+options nouveau modeset=0
```
Comment (t-nelson) [inline net/datacenter-node-install/scripts/disable-nouveau.sh]: Setting options on a blacklisted module? :thinking:

File: net/datacenter-node-install/scripts/setup-grub.sh [hunk 3]
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env bash
+
+if [ "$EUID" -ne 0 ]
+  then echo "Please run as root"
+  exit
+fi
+
+set -xe
+
+systemctl disable systemd-networkd-wait-online.service
```
Comment (t-nelson) [inline net/datacenter-node-install/scripts/setup-grub.sh]: Seems unrelated to grub :upside_down_face:

File: net/datacenter-node-install/scripts/setup-validator.sh [hunk 4]
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env bash
+
+if [ "$EUID" -ne 0 ]
+  then echo "Please run as root"
+  exit
+fi
+
+set -xe
+
+systemctl disable systemd-networkd-wait-online.service
```
Comment (t-nelson) [inline net/datacenter-node-install/scripts/setup-validator.sh]: Should these be calling the above sub-scripts to avoid maintenance burden?

File: net/datacenter-node-install/scripts/disable-nouveau.sh (hunk 2 above)
Comment (TristanDebrunner) [reply net/datacenter-node-install/scripts/disable-nouveau.sh]: well...it works the way it is

File: net/datacenter-node-install/setup-dc-node-1.sh [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+USERNAME="$(whoami)"
+
+set -xe
+
+apt update
+apt upgrade -y
+apt install -y build-essential pkg-config clang
+
+../scripts/install-docker.sh
```
Comment (danpaul000) [inline net/datacenter-node-install/setup-dc-node-1.sh]: Instead of relative paths, these should all be relative to the location of the running script, otherwise running them from some other dir other than what is assumed here will not work.

File: net/datacenter-node-install/setup-dc-node-1.sh [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+usermod -aG docker USERNAME
+
+../scripts/install-certbot.sh
+
```
Comment (danpaul000) [inline net/datacenter-node-install/setup-dc-node-1.sh]: rm whitespace between each command

File: net/datacenter-node-install/setup-dc-node-1.sh [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+./setup-sudoers.sh
+
+./setup-ssh.sh
+
+# Allow admin user to log in
+mkdir .ssh
```
Comment (danpaul000) [inline net/datacenter-node-install/setup-dc-node-1.sh]: ~/.ssh.   Always use explicit relative paths or absolute paths, never floating relative.

File: net/datacenter-node-install/setup-dc-node-1.sh [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+../scripts/install-redis.sh
+
+../scripts/install-rsync.sh
+
+../scripts/install-libssl-compatability.sh
+
+# Setup kernel constants
+cat > /etc/sysctl.d/20-solana-node.conf <<EOF
+
+# Solana networking requirements
```
Comment (danpaul000) [inline net/datacenter-node-install/setup-dc-node-1.sh]: Are these ever subject to change and therefore should they be popped out to a config file?

File: net/datacenter-node-install/setup-dc-node-2.sh [hunk 9]
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env bash
+
+if [[ "$EUID" -ne 0 ]]; then
+  echo "Please run as root"
+  exit 1
+fi
+
+set -xe
+
+./setup-cuda.sh
```
Comment (danpaul000) [inline net/datacenter-node-install/setup-dc-node-2.sh]: Make path explicit

File: net/datacenter-node-install/setup-dc-node-2.sh [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+# setup persistence mode across reboots
+tar -xvf /usr/share/doc/NVIDIA_GLX-1.0/sample/nvidia-persistenced-init.tar.bz2
+sudo ./nvidia-persistenced-init/install.sh systemd
```
Comment (danpaul000) [inline net/datacenter-node-install/setup-dc-node-2.sh]: explicit path

File: net/datacenter-node-install/setup-dc-node-2.sh (hunk 9 above)
Comment (danpaul000) [reply net/datacenter-node-install/setup-dc-node-2.sh]: At the file header, add an explicit path to the running script with something like:
`here=$(dirname "$0")`
Then later you can run:
`"$here"/setup-cuda.sh` or whatever, so you can call these scripts regardless of what your shell's PWD is.

File: net/datacenter-node-install/setup-dc-node-1.sh (hunk 8 above)
Comment (t-nelson) [reply net/datacenter-node-install/setup-dc-node-1.sh]: These come from https://github.com/solana-labs/solana/blob/8b7cca986a93ae2af24e2eda8e52a9df0581dff9/net/scripts/network-config.sh.  Looks like @pgarg66 might have the best answer?

**Comment by t-nelson** [comment]
If @TristanDebrunner doesn't beat me to it, I can take the external script path fixes over after I finish the allocation script

**Comment by danpaul000** [comment]
Can we also get a README.md that provides some top level instructions on which scripts to run, in which order, and under what circumstances?  Do these get run once ever, or all the time, or somewhere in between?

**Comment by t-nelson** [comment]
I believe they are a run once after OS install on nodes we host, type thing
 1) `setup-dc-node-1.sh`
 2) `reboot`
 3) `setup-dc-node-2.sh`

Assuming everything JustWorks(TM), that is...  I can add a simple `README.md` for now and let @TristanDebrunner add any details he's aware of when he has time.

**Comment by danpaul000** [comment]
Sounds good, thanks!

File: net/datacenter-node-install/setup-dc-node-1.sh [hunk 11]
```diff
@@ -0,0 +1,80 @@
+#!/usr/bin/env bash
+
+if [[ "$EUID" -ne 0 ]]; then
```
Comment (t-nelson) [inline net/datacenter-node-install/setup-dc-node-1.sh]: Something I noticed while doing the cleanup round...

Depending on how the user takes root, some of the variables we make assumptions about in this script may have unexpected values.  Affected are `USERNAME` (Both the normal shell global and the result of `whoami` used here) and `HOME`.

File: net/datacenter-node-install/setup-dc-node-1.sh [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+
+"$HERE"/../scripts/install-docker.sh
+usermod -aG docker "$USERNAME"
+"$HERE"/../scripts/install-certbot.sh
+"$HERE"/setup-sudoers.sh
+"$HERE"/setup-ssh.sh
+
+# Allow admin user to log in
+BASE_SSH_DIR="${HOME}/.ssh"
+mkdir "$BASE_SSH_DIR"
```
Comment (danpaul000) [inline net/datacenter-node-install/setup-dc-node-1.sh]: Won't $HOME be root's homedir in this case?  Looks like you want $USERNAME's homedir, no?
Can use
`su - ${USERNAME} -c 'mkdir "$BASE_SSH_DIR"'` and others, so you don't need to `chown` after every file/dir creation

File: net/datacenter-node-install/setup-dc-node-2.sh [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+if [[ "$EUID" -ne 0 ]]; then
+  echo "Please run as root"
+  exit 1
+fi
+
+HERE="$(dirname "$0")"
+
+set -xe
+
+"$HERE"/setup-cuda.sh
```
Comment (danpaul000) [inline net/datacenter-node-install/setup-dc-node-2.sh]: Do we always want to run this?  Our CPU nodes don't need cuda.  Can we check for the presence of certain cards before running this?

To make things complicated, there ARE graphics cards in the CPU nodes, but cheapo ones for local console access only.

File: net/datacenter-node-install/setup-validator.sh [hunk 14]
```diff
@@ -0,0 +1,21 @@
+#!/usr/bin/env bash
```
Comment (danpaul000) [inline net/datacenter-node-install/setup-validator.sh]: What is this script doing?  Looks like we call all those scripts already, and then randomize the password?  Why?

Comment (t-nelson) [reply net/datacenter-node-install/setup-validator.sh]: @TristanDebrunner would have the best answer here, but I believe this is a separate workflow (for setting up nodes we send to validators) than what we used to setup the nodes destined for our colo rack.

File: net/datacenter-node-install/setup-dc-node-2.sh (hunk 13 above)
Comment (t-nelson) [reply net/datacenter-node-install/setup-dc-node-2.sh]: My gut says, "it won't hurt," but I think this is another one we'd need an answer from @TristanDebrunner on.

We *could* parse the output of `lspci` to figure out if/what cards are installed

File: net/datacenter-node-install/setup-dc-node-1.sh (hunk 12 above)
Comment (t-nelson) [reply net/datacenter-node-install/setup-dc-node-1.sh]: I just did the obvious thing to communicate the intent of the change.  There are bigger problems, see https://github.com/solana-labs/solana/pull/5517#discussion_r315990902

File: net/datacenter-node-install/setup-dc-node-1.sh (hunk 11 above)
Comment (t-nelson) [reply net/datacenter-node-install/setup-dc-node-1.sh]: The most obvious solution I see here is to enforce that the scripts are called with root taken via `sudo` rather than from a root shell (ala. `sudo su --`).  Then we'll have the `SUDO_*` envvars to detect the enforcement *and* fill out all of our script-local variables.

How does that sound to you @danpaul000 ?

Comment (danpaul000) [reply net/datacenter-node-install/setup-dc-node-1.sh]: Yeah, I think it's reasonable to ask the user to just call `sudo this_script.sh`, as that is what the instruction is in the README, so now a future user has no excuse. :)

Comment (t-nelson) [reply net/datacenter-node-install/setup-dc-node-1.sh]: I'll put the checks in place and make the appropriate changes tomorrow

Comment (t-nelson) [reply net/datacenter-node-install/setup-dc-node-1.sh]: Addressed in https://github.com/solana-labs/solana/pull/5517/commits/85b7b72cfc0b597e13b4d533ab993e5d8d3ab862
